### PR TITLE
support multiple languages

### PR DESCRIPTION
**Pull request by harrisonpim**
- adds new language fields for titles and lettering
- runs a fuzzy match on existing title and lettering fields to handle transliteration issues

### CODE REVIEW

File: api/api/src/main/scala/uk/ac/wellcome/platform/api/elasticsearch/WorksQuery.scala [hunk 1]
```diff
@@ -17,6 +17,10 @@ case object WorksMultiMatcher {
   def apply(q: String): BoolQuery = {
     boolQuery()
       .should(
+        // we're running elasticsearch 7.9 at the moment - upgrading to 7.11 will give us the ability to run case
+        // insensitive prefixQueries, which will solve our "old and new london" problem
+        // (see https://github.com/wellcomecollection/catalogue/issues/1336)
+        prefixQuery("data.title.keyword", q).boost(1000),
```
Comment (jamesgorrie) [inline api/api/src/main/scala/uk/ac/wellcome/platform/api/elasticsearch/WorksQuery.scala:23]: @harrisonpim - we're on 7.11 now - do you want to add this to this PR?

File: api/api/src/main/scala/uk/ac/wellcome/platform/api/elasticsearch/WorksQuery.scala [hunk 2]
```diff
@@ -34,17 +38,25 @@ case object WorksMultiMatcher {
...
-          operator = Some(AND),
+          `type` = Some(BEST_FIELDS),
+          operator = Some(OR),
           fields = Seq(
-            (Some(1000), "data.contributors.agent.label"),
             (Some(100), "data.title"),
             (Some(100), "data.title.english"),
             (Some(100), "data.title.shingles"),
             (Some(100), "data.alternativeTitles"),
+            (None, "data.lettering"),
```
Comment (jamesgorrie) [inline api/api/src/main/scala/uk/ac/wellcome/platform/api/elasticsearch/WorksQuery.scala:50]: Why did lettering get in here?

File: api/api/src/main/scala/uk/ac/wellcome/platform/api/elasticsearch/WorksQuery.scala [hunk 3]
```diff
@@ -34,17 +38,25 @@ case object WorksMultiMatcher {
             (Some(1000), "data.imageData.id.otherIdentifiers.value"),
           ).map(f => FieldWithOptionalBoost(f._2, f._1.map(_.toDouble)))
         ),
-        prefixQuery("data.title.keyword", q).boost(1000),
         MultiMatchQuery(
           q,
-          `type` = Some(CROSS_FIELDS),
-          operator = Some(AND),
+          `type` = Some(BEST_FIELDS),
+          operator = Some(OR),
```
Comment (jamesgorrie) [inline api/api/src/main/scala/uk/ac/wellcome/platform/api/elasticsearch/WorksQuery.scala:44]: Almost 100% sure we don't want the `OR` query, especially boosted so highly.

e.g. `old and new london`

matches `old OR and OR new OR london` and boosts appropriately.

File: common/elasticsearch/src/main/scala/uk/ac/wellcome/elasticsearch/IndexConfig.scala [hunk 4]
```diff
@@ -41,6 +41,41 @@ trait IndexConfigFields {
...
+
+  def arabicTextField(name: String) =
+    textField(name).fields(
+      textField("arabic").analyzer("arabic")
+    )
+
+  // def persianTextField(name: String) =
+  //   textField(name).fields(
+  //     textField("persian").analyzer("persian")
+  //   )
```
Comment (jamesgorrie) [inline common/elasticsearch/src/main/scala/uk/ac/wellcome/elasticsearch/IndexConfig.scala:72]: Could we just remove this for now?

Comment (jamesgorrie) [reply common/elasticsearch/src/main/scala/uk/ac/wellcome/elasticsearch/IndexConfig.scala:72]: Or leave a comment / issue to fix it as Persian is pretty important to us isn't it?

File: api/api/src/main/scala/uk/ac/wellcome/platform/api/elasticsearch/WorksQuery.scala (hunk 2 above)
Comment (harrisonpim) [reply api/api/src/main/scala/uk/ac/wellcome/platform/api/elasticsearch/WorksQuery.scala:50]: lettering is another field where we see multiple languages used, so imo we should be applying fuzz to it to catch transliteration issues

Comment (jamesgorrie) [reply api/api/src/main/scala/uk/ac/wellcome/platform/api/elasticsearch/WorksQuery.scala:50]: Assuming this is language type fields, maybe we could look at this implementation once we search multiple language fields?

Comment (jamesgorrie) [reply api/api/src/main/scala/uk/ac/wellcome/platform/api/elasticsearch/WorksQuery.scala:50]: I wasn't very clear, I wonder if we could leave this implementation out until we actually search across multiple language fields.

File: api/api/src/main/scala/uk/ac/wellcome/platform/api/elasticsearch/WorksQuery.scala (hunk 1 above)
Comment (harrisonpim) [reply api/api/src/main/scala/uk/ac/wellcome/platform/api/elasticsearch/WorksQuery.scala:23]: yep

File: api/api/src/main/scala/uk/ac/wellcome/platform/api/elasticsearch/WorksQuery.scala (hunk 2 above)
Comment (harrisonpim) [reply api/api/src/main/scala/uk/ac/wellcome/platform/api/elasticsearch/WorksQuery.scala:50]: sorry, i should have specified - lettering is a place where we also see _transliteration_ issues (not just multiple languages), which is what the fuzzy match is intended to solve

Comment (jamesgorrie) [reply api/api/src/main/scala/uk/ac/wellcome/platform/api/elasticsearch/WorksQuery.scala:50]: What is the point of the `BEST_FIELDS` then?

File: api/api/src/main/scala/uk/ac/wellcome/platform/api/elasticsearch/WorksQuery.scala [hunk 5]
```diff
@@ -34,17 +38,25 @@ case object WorksMultiMatcher {
...
+          `type` = Some(BEST_FIELDS),
           operator = Some(AND),
           fields = Seq(
-            (Some(1000), "data.contributors.agent.label"),
             (Some(100), "data.title"),
             (Some(100), "data.title.english"),
             (Some(100), "data.title.shingles"),
             (Some(100), "data.alternativeTitles"),
+            (None, "data.lettering"),
+          ).map(f => FieldWithOptionalBoost(f._2, f._1.map(_.toDouble))),
```
Comment (alexwlchan) [inline api/api/src/main/scala/uk/ac/wellcome/platform/api/elasticsearch/WorksQuery.scala:51]: Maybe use named variables here? Something like:

```suggestion
          ).map( case (weight, fieldName) => FieldWithOptionalBoost(fieldname, weight.map(_.toDouble))),
```

File: common/elasticsearch/src/main/scala/uk/ac/wellcome/elasticsearch/WorksAnalysis.scala [hunk 6]
```diff
@@ -46,9 +59,65 @@ object WorksAnalysis {
...
+
+  val frenchAnalyzer = CustomAnalyzer(
+    "french_analyzer",
+    tokenizer = "standard",
+    tokenFilters = List(
+      "lowercase",
+      frenchStemmerTokenFilter.name
+    ),
+    charFilters = Nil
+  )
```
Comment (alexwlchan) [inline common/elasticsearch/src/main/scala/uk/ac/wellcome/elasticsearch/WorksAnalysis.scala:76]: Having a function to create these CustomAnalyzers for each language would reduce the repetition, and make it more obvious if they're all the same or differ in slight ways.

File: common/elasticsearch/src/main/scala/uk/ac/wellcome/elasticsearch/IndexConfig.scala [hunk 7]
```diff
@@ -41,6 +41,36 @@ trait IndexConfigFields {
       textField("english").analyzer("english")
     )
 
+  def frenchTextField(name: String) =
+    textField(name).fields(
+      textField("french").analyzer("french")
+    )
```
Comment (alexwlchan) [inline common/elasticsearch/src/main/scala/uk/ac/wellcome/elasticsearch/IndexConfig.scala:47]: Possibly a daft question, but what are these functions for? I don't see you using them anywhere.

**Comment by harrisonpim** [comment]
thanks for all this tidying up @jamesgorrie! all looking much clearer 😁

File: api/api/src/main/scala/uk/ac/wellcome/platform/api/elasticsearch/WorksQuery.scala (hunk 1 above)
Comment (jamesgorrie) [reply api/api/src/main/scala/uk/ac/wellcome/platform/api/elasticsearch/WorksQuery.scala:23]: To keep a record of our conversation, elastic4s 7.10 doesn't have this added.

File: common/elasticsearch/src/main/scala/uk/ac/wellcome/elasticsearch/WorksAnalysis.scala (hunk 6 above)
Comment (jamesgorrie) [reply common/elasticsearch/src/main/scala/uk/ac/wellcome/elasticsearch/WorksAnalysis.scala]: They've been removed now...

File: common/elasticsearch/src/main/scala/uk/ac/wellcome/elasticsearch/IndexConfig.scala (hunk 7 above)
Comment (jamesgorrie) [reply common/elasticsearch/src/main/scala/uk/ac/wellcome/elasticsearch/IndexConfig.scala]: They've been removed now...

**Review by jamesgorrie — APPROVED** [review]

**Pull request merged**